shared/src/main/kotlin/org/javacs/kt/storage/Storage.kt

```diff
@@ -0,0 +1,74 @@
+package org.javacs.kt.storage
+
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.builtins.SetSerializer
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
+import kotlinx.serialization.encoding.Decoder
+import kotlinx.serialization.encoding.Encoder
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.serializer
+import org.javacs.kt.LOG
+import java.lang.Exception
+import java.nio.file.FileAlreadyExistsException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+
+class Storage(val path: Path) {
+    fun getSlice(slicePath: String): Storage? {
+        val fullSlicePath = Paths.get(path.toString(), slicePath)
+
+        try {
+            Files.createDirectories(fullSlicePath)
+        } catch (ex: FileAlreadyExistsException) {
+            return null
+        }
+
+        return Storage(fullSlicePath)
+    }
+
+    inline fun <reified T> getObject(objectPath: String, deserializer: KSerializer<T> = Json.serializersModule.serializer()): T? {
+        val fullObjectPath = Paths.get(path.toString(), objectPath)
+
+        if (Files.exists(fullObjectPath)) {
+            val content = fullObjectPath.toFile().readText()
+            try {
+                return Json.decodeFromString(deserializer, content)
+            } catch (ex: Exception) {
+                // We catch any exception just in case something unexpected happens. We return null in that case and delete the file.
+                LOG.printStackTrace(ex)
+                Files.deleteIfExists(fullObjectPath)
+            }
+        }
+
+        return null
+    }
+
+    inline fun <reified T> setObject(objectPath: String, value: T, serializer: KSerializer<T> = Json.serializersModule.serializer()) {
+        val fullObjectPath = Paths.get(path.toString(), objectPath)
+
+        if (!Files.exists(fullObjectPath)) {
+            Files.createFile(fullObjectPath)
+        }
```

Comment:
Just an idea, could we use Kotlin's subscript overloading here, i.e. make `getObject` and `setObject` `operator fun`s (see [the docs](https://kotlinlang.org/docs/operator-overloading.html#unary-operations))? Not sure whether that works with the optional serializer parameters.

Reply:
Yeah, this would be really cool, but the optional serializer parameters kind of ruin it unfortunately. We also have the storage as a nullable property everywhere, so I don't think we'd get that much out of it even then (as far as I know, the subscript operator doesn't work well with the elvis operator for example).